src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5246 @@
+                
+                if (tag.TagClass != TagClass.Universal)
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                totalRead += headerLength;
+
+                if (isIndefinite && tag.TagValue == (int)UniversalTagNumber.EndOfContents)
+                {
+                    ValidateEndOfContents(tag, length, headerLength);
+
+                    bytesRead = totalRead;
+                    return totalContent;
+                }
+
+                if (tag.TagValue != (int)UniversalTagNumber.OctetString)
+                {
+                    // T-REC-X.690-201508 sec 8.7.3.2 (in particular, Note 2)
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                if (ruleSet == AsnEncodingRules.CER)
+                {
+                    if (tag.IsConstructed || lastSegmentLength != MaxCERSegmentSize)
+                    {
+                        // T-REC-X.690-201508 sec 9.2
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+                }
+
+                cur = cur.Slice(headerLength);
+
+                if (length == null)
+                {
+                    totalContent += CopyConstructedOctetString(
+                        cur,
+                        ref destination,
+                        write,
+                        ruleSet,
+                        true,
+                        ref lastSegmentLength,
+                        out int nestedBytesRead);
+
+                    totalRead += nestedBytesRead;
+                    cur = cur.Slice(nestedBytesRead);
+                }
+                else if (tag.IsConstructed)
+                {
+                    totalContent += CopyConstructedOctetString(
+                        Slice(cur, 0, length.Value),
+                        ref destination,
+                        write,
+                        ruleSet,
+                        false,
+                        ref lastSegmentLength,
+                        out int nestedContentRead);
+
+                    totalRead += nestedContentRead;
+                    cur = cur.Slice(nestedContentRead);
+                }
+                else
+                {
+                    int lengthValue = length.Value;
+
+                    // T-REC-X.690-201508 sec 9.2
+                    if (ruleSet == AsnEncodingRules.CER && lengthValue > MaxCERSegmentSize)
+                    {
+                        throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                    }
+
+                    totalRead += lengthValue;
+                    totalContent += lengthValue;
+                    lastSegmentLength = lengthValue;
+
+                    ReadOnlyMemory<byte> segment = Slice(cur, 0, lengthValue);
+             
```

Comment:
Even though this is a Try method, it throws an exception if the internal object state is invalid. We should document this behavior since callers might not expect it.